apps/nouns-camp/src/hooks/simulation.js

```diff
@@ -0,0 +1,103 @@
+import React from "react";
+import { decodeAbiParameters, encodeFunctionData, parseAbiItem } from "viem";
+import { unparse } from "../utils/transactions";
+
+const SIMULATE_TRANSACTION_TYPES = [
+  "function-call",
+  "proxied-function-call",
+  "payable-function-call",
+  "proxied-payable-function-call",
+  "treasury-noun-transfer",
+  "escrow-noun-transfer",
+  "transfer",
+  "usdc-transfer-via-payer",
+];
+
+export const fetchSimulation = async ({
+  target,
+  value,
+  signature,
+  calldata,
+}) => {
+  if (!signature) signature = "transfer()";
+
+  const { name, inputs } = parseAbiItem(`function ${signature}`);
+  const args = decodeAbiParameters(inputs, calldata);
+
+  const encodedData = encodeFunctionData({
+    abi: [
+      {
+        inputs: inputs,
+        name: name,
+        type: "function",
+      },
+    ],
+    functionName: name,
+    args: args,
+  });
+
+  const parsedTransaction = {
+    to: target,
+    value: value || "0",
+    input: encodedData,
+  };
+
+  const body = JSON.stringify(parsedTransaction);
+  const res = await fetch("/api/simulate", {
+    method: "POST",
+    headers: {
+      "Content-Type": "application/json",
+    },
+    body,
+  });
+
+  if (!res.ok) {
+    throw new Error("Failed to simulate transaction");
+  }
+
+  const simulation = await res.json();
+  return simulation;
+};
+
+export const useTransactionSimulation = (transaction) => {
+  const [isFetching, setIsFetching] = React.useState(false);
+  const [simulation, setSimulation] = React.useState(null);
+
+  const fetchData = React.useCallback(async () => {
+    try {
+      setIsFetching(true);
+      if (!SIMULATE_TRANSACTION_TYPES.includes(transaction.type)) {
+        setSimulation(null);
+        return;
+      }
+
+      const { targets, values, signatures, calldatas } = unparse([transaction]);
+
+      const tx = {
```

Comment:
is there a prettier JS way of doing this assignment?

Reply:
truth is in the near future i'll add multiple tx simulation so maybe not a big issue rn

Reply:
Not really!